### PR TITLE
[Emails] Prevent changes to emails in specific states

### LINKED ISSUE #2459: [Emails] Prevent changes to scheduled emails in specific states
As a system administrator, I want users to not be able to cancel a successfully run scheduled email, or I don't want them to reschedule a cancelled email, or I don't want to allow any changes to a locked email.

There may be more forbidden transitions than these to.

### PR DESCRIPTION
**Pull request by pbanaszkiewicz**
Fixes #2459.

Based on PR #2491.

This PR introduces whitelist of statuses the scheduled email has to be in in order to edit, reschedule or cancel it.

One of the interesting techniques used here is database-level locking connected with filtering for allowed statuses. This will ensure that an email is in proper status before and during the POST request.

### CODE REVIEW

**Comment by pbanaszkiewicz** [comment]
@elichad This is ready for your review.

**Comment by elichad** [comment]
@pbanaszkiewicz this PR contains a lot of the commits from #2491, could you rebase it?

**Comment by pbanaszkiewicz** [comment]
@elichad Whoops, I completely missed that. Rebased and ready for review. Thanks!

**Comment by pbanaszkiewicz** [comment]
@elichad Can you review this PR? Thanks in advance!

File: amy/templates/emails/scheduled_email_reschedule.html
Comment (elichad) [inline amy/templates/emails/scheduled_email_reschedule.html:18]:
```suggestion
  <strong>Warning</strong>: if you reschedule a cancelled email, it will become scheduled again and will be sent immediately if the scheduled time is in the past.
```

File: amy/emails/views.py
Comment (elichad) [inline amy/emails/views.py:165]: I wondered if this would lock the object as soon as the user navigates to the Update page, but this appears not to be the case (I assume I misunderstood how `select_for_update()` works):
In one tab I opened the Update view for a scheduled email, then in another tab I cancelled that email. When I submitted the update, I got a 404 error because the email can no longer be edited.

Could we add a check during the form validation stage for these views that shows an error to the user if the object no longer has the right status for the operation? (can be a separate issue)

**Review by elichad — APPROVED** [review]
Sorry for the long delay. This looks good to me! I've added a couple of minor comments and one suggestion for a new issue

File: amy/emails/models.py
Comment (pbanaszkiewicz) [reply amy/emails/models.py:129]: I admit it's debatable. I wanted to not prevent user from amending the email just because I may not see the use case they may have.

@maneesha Once email is sent successfully, should users be able to edit it? This will not affect the email whatsoever, including re-sending it.

File: amy/emails/tests/test_controller.py
Comment (pbanaszkiewicz) [reply amy/emails/tests/test_controller.py:223]: Fixed, thanks!

File: amy/emails/tests/test_views.py
Comment (pbanaszkiewicz) [reply amy/emails/tests/test_views.py:469]: Good catch, I'm fixing these mistakes.

File: amy/templates/emails/scheduled_email_reschedule.html
Comment (pbanaszkiewicz) [reply amy/templates/emails/scheduled_email_reschedule.html:18]: I'll use your suggestion, but it doesn't capture exactly what I had in my mind. The email can be sent (almost) immediately if the scheduled time is in the past or very close to current time.

File: amy/emails/views.py
Comment (pbanaszkiewicz) [reply amy/emails/views.py:165]: `select_for_update()` is a Django way of creating a `SELECT ... FOR UPDATE` PostgreSQL query. It's one of the methods of per-row explicit locking: https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS

This locks the row for the duration of a single transaction, and I wanted to use this because there's a risk someone will be updating an entry that simultaneously was fetched by the email worker and made "locked" (as in the scheduled email status).

You will have hard time reproducing this one on your local machine, the best way is to have a command that spawns multiple clients trying to do one single edit. I've had some very bad experience with this kind of behavior which resulted in a race condition that was difficult to catch.


> Could we add a check during the form validation stage for these views that shows an error to the user if the object no longer has the right status for the operation? (can be a separate issue)

That is possible, but I don't know how (yet). I'll add a low-priority issue for this.

Comment (pbanaszkiewicz) [reply amy/emails/views.py:165]: Ticket created: https://github.com/carpentries/amy/issues/2531

**Comment by pbanaszkiewicz** [comment]
@elichad updated, now I'm waiting for feedback from @maneesha before merging.

File: amy/templates/emails/scheduled_email_reschedule.html
Comment (elichad) [reply amy/templates/emails/scheduled_email_reschedule.html]: Makes sense. I was thinking only about capturing the part of the behaviour that might be unexpected

File: amy/emails/models.py
Comment (pbanaszkiewicz) [reply amy/emails/models.py:129]: @maneesha decided to block editing.

Comment (maneesha) [reply amy/emails/models.py:129]: We should actually have enforcement that an email **can not** be edited after it has been sent, as this would create a situation where what is in AMY is different than what was actually set.

**Pull request merged**
